prow/tide/contextregister.go

```diff
@@ -0,0 +1,127 @@
+/*
+Copyright 2017 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+// Package tide contains a controller for managing a tide pool of PRs. The
+// controller will automatically retest PRs in the pool and merge them if they
+// pass tests.
+
+package tide
+
+import (
+	"sync"
+
+	"github.com/shurcooL/githubql"
+	"k8s.io/apimachinery/pkg/util/sets"
+	"k8s.io/test-infra/prow/config"
+)
+
+// ignoreContexts abstracts checking for optional contexts for a PR.
+type ignoreContexts interface {
+	ignoreContext(string) bool
+	getRequiredContexts() sets.String
+}
+
+func newContext(c string) Context {
+	return Context{
+		Context:     githubql.String(c),
+		State:       githubql.StatusStateExpected,
+		Description: githubql.String(""),
+	}
+}
+
+// tideIC ignore tide context.
+type tideIC struct{}
+
+// contextRegister contains all registered ignoreContexts.
+type contextRegister struct {
+	lock sync.RWMutex
+	ics  []ignoreContexts
+}
+
+// branchProtectionIC implements ignoreContexts for the branch protection plugin.
+type branchProtectionIC struct {
+	requiredContexts sets.String
+}
+
+// newContextRegister should be instantiated in tide for each combination of org, repo, branch.
+// It contains all the registered ignoreContexts
+func newContextRegister() *contextRegister {
```

Comment:
Can you explain why `newContextRegister` and `ic.Register()` are necessary?

Reply:
so basically we have a default register which is composed of tideIC ignoreContext interface implementation, because we always want to skip tide context. When branch protection is enabled in tide, we register the branch protection context.